### PR TITLE
PP-3680 Fixed url host in links to reference public api

### PR DESCRIPTION
**Pull request by danworth**
- Created a DirectDebitEventsService since creation of response is now a little
more complex.
- Factored out calling dd connector and processing the response into the new
service.
- Added method to PublicApiUrlGenerator to change the host of a link to publicAPI.
- Updated tests to confirm host on links.
- Updated DirectDebitEventResponse to add neccessary getters and setters to
manipulate the links as above.
- Corrected how the agreement and payment links are created for each event
to include the host. This is done by using the existing PublicApiUriGenerator
class.
- Updated the tests to confirm agreement and payment links set correctly.
- Deleted the DirectDebitEventTest which was checking that DirectDebitEvent was
correctly setting the links, however this operation is now done by the new
DirectDebitEventService and so no longer appropriate to test here.

### CODE REVIEW

File: src/test/java/uk/gov/pay/api/it/GetDirectDebitEventsTest.java [hunk 1]
```diff
@@ -65,13 +61,13 @@ public void getDirectDebitEvents() {
                 .body("results[0].event_date", is("2018-03-13T10:00:04.666Z"))
                 .body("results[0].event", is("PAYMENT_ACKNOWLEDGED_BY_PROVIDER"))
                 .body("results[0].event_type", is("CHARGE"))
-                .body("results[0]._links.agreement", is("/v1/agreements/1"))
-                .body("results[0]._links.payment", is("/v1/payments/4"))
+                .body("results[0]._links.agreement", is(app.getConfiguration().getBaseUrl() + "v1/agreements/1"))
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/api/it/GetDirectDebitEventsTest.java]: Probably is simpler to see what value we test by extracting `app.getConfiguration.getBaseUrl()` into a local variable and reuse it

File: src/test/java/uk/gov/pay/api/it/GetDirectDebitEventsTest.java [hunk 2]
```diff
@@ -65,13 +61,13 @@ public void getDirectDebitEvents() {
...
+                .body("results[0]._links.payment", is(app.getConfiguration().getBaseUrl() + "v1/payments/4"))
                 .body("_links.next_page", isEmptyOrNullString())
                 .body("_links.prev_page", isEmptyOrNullString())
-                .body("_links.self.href", is("/v1/events?to_date=2018-03-13T10:00:04Z&from_date=2018-03-13T10:00:04Z&agreement_external_id=1&page=1&display_size=100"))
-                .body("_links.last_page.href", is("/v1/events?to_date=2018-03-13T10:00:04Z&from_date=2018-03-13T10:00:04Z&agreement_external_id=1&page=1&display_size=100"))
-                .body("_links.first_page.href", is("/v1/events?to_date=2018-03-13T10:00:04Z&from_date=2018-03-13T10:00:04Z&agreement_external_id=1&page=1&display_size=100"));
+                .body("_links.self.href", is( app.getConfiguration().getBaseUrl() + "v1/events?to_date=2018-03-13T10%3A00%3A04Z&from_date=2018-03-13T10%3A00%3A04Z&agreement_external_id=1&page=1&display_size=100"))
```
[the remark this answers is not in the document]
Comment (danworth) [reply src/test/java/uk/gov/pay/api/it/GetDirectDebitEventsTest.java]: This is being caused by URIBuilder.replaceQuery() which encodes the query string prior to updating. Therefore when we call toString() the query is encoded (":" = %3A). As discussed I've now added a step to decode the URL.

File: src/test/java/uk/gov/pay/api/it/GetDirectDebitEventsTest.java (hunk 1 above)
Comment (danworth) [reply src/test/java/uk/gov/pay/api/it/GetDirectDebitEventsTest.java]: Agreed, I've made this change.

File: src/main/java/uk/gov/pay/api/resources/DirectDebitEventsResource.java [hunk 3]
```diff
@@ -7,11 +7,11 @@
 import io.swagger.annotations.ApiParam;
 import io.swagger.annotations.ApiResponse;
 import io.swagger.annotations.ApiResponses;
-import org.apache.http.HttpStatus;
 import uk.gov.pay.api.auth.Account;
 import uk.gov.pay.api.exception.ConnectorResponseErrorException;
```
[the remark this answers is not in the document]
Comment (danworth) [reply src/main/java/uk/gov/pay/api/resources/DirectDebitEventsResource.java]: Thanks

File: src/main/java/uk/gov/pay/api/resources/DirectDebitEventsResource.java [hunk 4]
```diff
@@ -37,11 +33,13 @@
 
     private final Client client;
```
[the remark this answers is not in the document]
Comment (danworth) [reply src/main/java/uk/gov/pay/api/resources/DirectDebitEventsResource.java]: Yep, sorry left over from the refactor.

**Pull request merged**